File: ios-sdk/Sources/MxLSdk/Session/SessionManager.swift
import Foundation

/**
 * Manages user sessions and context.
 */
internal class SessionManager {
    private let configuration: SdkConfiguration
    private let userDefaults: UserDefaults
    
    private var currentSessionId: String?
    private var sessionStartTime: TimeInterval = 0
    private var lastActivityTime: TimeInterval = 0
    
    private let sessionTimeout: TimeInterval = 30 * 60 // 30 minutes
    
    private enum Keys {
        static let sessionId = "mxl_session_id"
        static let userId = "mxl_user_id"
    }
    
    static func create(configuration: SdkConfiguration) -> SessionManager {
        return SessionManager(configuration: configuration)
    }
    
    private init(configuration: SdkConfiguration) {
        self.configuration = configuration
        self.userDefaults = UserDefaults.standard
        startNewSession()
    }
    
    /**
     * Start a new session.
     */
    func startNewSession() {
        let now = Date().timeIntervalSince1970
        
        // Check if current session has expired
        if let sessionId = currentSessionId,
           (now - lastActivityTime) > sessionTimeout {
            endSession()
        }
        
        if currentSessionId == nil {
            currentSessionId = UUID().uuidString
            sessionStartTime = now
            lastActivityTime = now
            
            userDefaults.set(currentSessionId, forKey: Keys.sessionId)
        }
    }
    
    /**
     * End the current session.
     */
    func endSession() {
        currentSessionId = nil
        sessionStartTime = 0
        lastActivityTime = 0
        userDefaults.removeObject(forKey: Keys.sessionId)
    }
    
    /**
     * Get the current session ID.
     */
    func getSessionId() -> String? {
        return currentSessionId
    }
    
    /**
     * Identify a user.
     */
    func identify(userId: String, attributes: [String: Any] = [:]) {
        userDefaults.set(userId, forKey: Keys.userId)
        // TODO: Store user attributes
    }
    
    /**
     * Get the current user ID.
     */
    func getUserId() -> String? {
        return userDefaults.string(forKey: Keys.userId)
    }
    
    /**
     * Called when app comes to foreground.
     */
    func onAppForegrounded() {
        startNewSession()
    }
    
    /**
     * Called when app goes to background.
     */
    func onAppBackgrounded() {
        lastActivityTime = Date().timeIntervalSince1970
    }
}

File: android-sdk/sdk/src/main/java/com/adlcom/mxl/sdk/session/SessionManager.kt
package com.adlcom.mxl.sdk.session

import android.app.Activity
import android.content.Context
import android.content.SharedPreferences
import com.adlcom.mxl.sdk.config.SdkConfiguration
import java.util.UUID

/**
 * Manages user sessions and context.
 */
class SessionManager private constructor(
    private val context: Context,
    private val configuration: SdkConfiguration,
    private val prefs: SharedPreferences
) {
    companion object {
        private const val PREFS_NAME = "mxl_sdk_prefs"
        private const val KEY_SESSION_ID = "session_id"
        private const val KEY_USER_ID = "user_id"
        private const val SESSION_TIMEOUT_MS = 30 * 60 * 1000L // 30 minutes
        
        fun create(context: Context, configuration: SdkConfiguration): SessionManager {
            val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            return SessionManager(context, configuration, prefs)
        }
    }
    
    private var currentSessionId: String? = null
    private var sessionStartTime: Long = 0
    private var lastActivityTime: Long = 0
    
    init {
        startNewSession()
    }
    
    /**
     * Start a new session.
     */
    fun startNewSession() {
        val now = System.currentTimeMillis()
        
        // Check if current session has expired
        if (currentSessionId != null && (now - lastActivityTime) > SESSION_TIMEOUT_MS) {
            endSession()
        }
        
        if (currentSessionId == null) {
            currentSessionId = UUID.randomUUID().toString()
            sessionStartTime = now
            lastActivityTime = now
            
            prefs.edit()
                .putString(KEY_SESSION_ID, currentSessionId)
                .apply()
        }
    }
    
    /**
     * End the current session.
     */
    fun endSession() {
        currentSessionId = null
        sessionStartTime = 0
        lastActivityTime = 0
    }
    
    /**
     * Get the current session ID.
     */
    fun getSessionId(): String? = currentSessionId
    
    /**
     * Identify a user.
     */
    fun identify(userId: String, attributes: Map<String, Any> = emptyMap()) {
        prefs.edit()
            .putString(KEY_USER_ID, userId)
            .apply()
        
        // Store user attributes
        // TODO: Store attributes in database
    }
    
    /**
     * Get the current user ID.
     */
    fun getUserId(): String? = prefs.getString(KEY_USER_ID, null)
    
    /**
     * Called when app comes to foreground.
     */
    fun onAppForegrounded() {
        startNewSession()
    }
    
    /**
     * Called when app goes to background.
     */
    fun onAppBackgrounded() {
        lastActivityTime = System.currentTimeMillis()
    }
    
    /**
     * Called when an activity is resumed.
     */
    fun onActivityResumed(activity: Activity) {
        lastActivityTime = System.currentTimeMillis()
        // TODO: Track screen view
    }
    
    /**
     * Called when an activity is paused.
     */
    fun onActivityPaused(activity: Activity) {
        lastActivityTime = System.currentTimeMillis()
    }
}

